kong/plugins/grpc-web/deco.lua

```diff
@@ -0,0 +1,212 @@
+local encode_json = cjson.encode
+local decode_json = cjson.decode
+
+local deco = {}
+deco.__index = deco
+
+ffi.cdef [[
+  typedef struct __attribute__ ((__packed__)) frame_hdr {
+    uint8_t frametype;
+    uint32_t be_framesize;
+  } frame_hdr;
+]]
+
+local frame_hdr = ffi.typeof("frame_hdr")
+local frame_hdr_p = ffi.typeof("const frame_hdr *")
+local HEADER_SIZE = ffi.sizeof("frame_hdr")
+
+
+local text_encoding_from_mime = {
+  ["application/grpc-web"] = "plain",
+  ["application/grpc-web-text"] = "base64",
+  ["application/grpc-web+proto"] = "plain",
+  ["application/grpc-web-text+proto"] = "base64",
+  ["application/grpc-web+json"] = "plain",
+  ["application/grpc-web-text+json"] = "base64",
+  ["application/json"] = "plain",
+}
+
+local framing_form_mime = {
+  ["application/grpc-web"] = "grpc",
+  ["application/grpc-web-text"] = "grpc",
+  ["application/grpc-web+proto"] = "grpc",
+  ["application/grpc-web-text+proto"] = "grpc",
+  ["application/grpc-web+json"] = "grpc",
+  ["application/grpc-web-text+json"] = "grpc",
+  ["application/json"] = "none",
+}
+
+local msg_encodign_from_mime = {
+  ["application/grpc-web"] = "proto",
+  ["application/grpc-web-text"] = "proto",
+  ["application/grpc-web+proto"] = "proto",
+  ["application/grpc-web-text+proto"] = "proto",
+  ["application/grpc-web+json"] = "json",
+  ["application/grpc-web-text+json"] = "json",
+  ["application/json"] = "json",
+}
+
+
+
+-- parse, compile and load .proto file
+-- returns the parsed info as a table
+local _proto_info = {}
+local function get_proto_info(fname)
+  local info = _proto_info[fname]
+  if info then
+    return info
+  end
+
+  local p = protoc.new()
+  local parsed = p:parsefile(fname)
+
+  info = {}
+
+  for _, srvc in ipairs(parsed.service) do
+    for _, mthd in ipairs(srvc.method) do
+      info[("/%s.%s/%s"):format(parsed.package, srvc.name, mthd.name)] = {
+        mthd.input_type, mthd.output_type
+      }
+    end
+  end
+
+  _proto_info[fname] = info
+
+  p:loadfile(fname)
+  return info
+end
+
+-- return input and output names of the method specified by the url path
+-- TODO: memoize
+local function rpc_types(path, protofile)
+  if not protofile then
+    return nil
+  end
+
+  local info = get_proto_info(protofile)
+  local types = info[path]
+  if not types then
+    return nil, ("Unkown path %q"):format(path)
+  end
+
+  return types[1], types[2]
+end
+
+
+function deco.new(mimetype, path, protofile)
+  local text_encoding = text_encoding_from_mime[mimetype]
+  local framing = framing_form_mime[mimetype]
+  local msg_encoding = msg_encodign_from_mime[mimetype]
+
+  local input_type, output_type
+  if msg_encoding ~= "proto" then
+    if not protofile then
+      return nil, "transcoding requests require a .proto file defining the service"
+    end
+    input_type, output_type = rpc_types(path, protofile)
+    if input_type == nil and output_type ~= nil then
+      return nil, output_type
+    end
+  end
+
+  return setmetatable({
```

Comment:
Should cache `setmetatable` at module level.

Also, we usually reuse the same metatable and save it at module level instead of creating a new table everytime.

Reply:
?
the metatable is the module...

Reply:
My apologize, I mistaken the actual table instance for the metatable. Yes this looks all fine to me.

One thing to mention is that since the `deco` is created a lot, https://github.com/openresty/lua-tablepool provides a simple interface for recycling the table instances so it doesn't have to be allocated every time.